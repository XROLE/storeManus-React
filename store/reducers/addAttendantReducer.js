import {
  ADDING_ATTENDANT,
  ADDED_ATTENDANT,
  ADD_ATTENDANT_ERROR,
} from '../actions/actionTypes';

const initialState = {
  message: '',
  pending: false,
  error: '',
};

export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_ATTENDANT_ERROR: {
      return {
        ...state,
        error: action.error,
        pending: false,
      };
    }
    case ADDED_ATTENDANT: {
      return {
        ...state,
        message: action.payload.message,
        pending: false,
      };
    }
    case ADDING_ATTENDANT: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }
    default:
      return state;
  }
}
