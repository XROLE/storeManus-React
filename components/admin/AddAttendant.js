import React, { Fragment, Component } from 'react';
import { ToastContainer, toast } from 'react-toastify';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import { Link } from 'react-router-dom';
import { addAttendant } from '../../store/reducers/loginReducer';
import Navbar from '../navbar/Navbar';
import 'react-toastify/dist/ReactToastify.css';

class AdminAddAttendant extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      shouldShowError: false,
      showSuccessMessage: false,
    };
  }

  onChange(e) {
    return this.setState({
      [e.target.name]: e.target.value,
    });
  }

  handleSubmit(e) {
    e.preventDefault();
    const attendantDetails = this.state;
    const { registerAttendant } = this.props;

    this.setState({ shouldShowError: true });
    registerAttendant(addAttendant(attendantDetails));
  }

  notify(error) {
    toast.error(error);
  }

  render() {
    const {
      email,
      firstName,
      lastName,
      shouldShowError,
      showSuccessMessage,
    } = this.state;
    const { registrationError, pending, message } = this.props;

    if (registrationError && shouldShowError) {
      this.notify(registrationError);
      this.setState({
        shouldShowError: false,
      });
    }
    if (showSuccessMessage && message) {
      console.log(message);
      this.setState({
        showSuccessMessage: false,
      });
    }

    return (
      <Fragment>
        <Navbar />
        <div className="dasboardDiv">
          <div className="dasboard-content">
            <div className="dashboard-toggler-div">
              <p className="toggler">
                <i className="fa fa-bars hanburgerButton" />
                <i className="fa fa-times close" />
              </p>
            </div>
            <div className="dasboard-content-main">
              <ul className="dasboard-content-items">
                <p className="dasboard-header"><Link to="/adminDashboard">Dashboard</Link></p>
                <li className="link">
                  <a href="admin-all-products.html">
                    <i className="fas fa-cookie-bite" />
                    &nbsp; All Products
                  </a>
                </li>
                <li className="link">
                  <a href="admin-available-products.html">
                    <i className="fab fa-accessible-icon" />
                    &nbsp; Available Products
                  </a>
                </li>
                <li className="link">
                  <a href="admin-finished-products.html">
                    <i className="fas fa-thermometer-empty" />
                    &nbsp; Finished Products
                  </a>
                </li>
                <li className="link">
                  <a href="admin-add-products.html">
                    <i className="far fa-plus-square" />
                    &nbsp; Add Product
                  </a>
                </li>
              </ul>
              <ul className="dasboard-content-items">
                <p className="dasboard-header">Attendants</p>
                <li className="link">
                  <a href="admin-all-attendants.html">
                    <i className="fas fa-users" />
                    &nbsp; All Attendants
                  </a>
                </li>
                <li className="link">
                  <Link to="/admin/add/attendant">
                    <i className="far fa-plus-square" />
                    &nbsp; Add Attendant
                  </Link>
                </li>
              </ul>
            </div>
          </div>
          <div className="test">
            <div className="avatar-div-container">
              <div className="avater-div">
                <img src="../../src/img/avatar.jpg" alt="Avatar" className="Avatar" />
                <p className="avatar-label fontBlack">welcome: Xrole</p>
              </div>
              <div className="filter-div">
                <label htmlFor="Filter sales by Attendants">
                  <a href="admin-sales-viewer.html">Filter Sales By: </a>
                </label>
                &nbsp;
                <select name="" id="">
                  <option value="">-Select All-</option>
                  <option value="">Attendats</option>
                  <option value="">Categories</option>
                  <option value="">Date</option>
                </select>
                &nbsp;
                <select name="" id="">
                  <option value="">---- Select All ----</option>
                  <option value="Cythia Paul">Cythia Paul</option>
                  <option value="Blessing Uche">Blessing Uche</option>
                  <option value="Silas Nwokeocha">Silas Nwokeocha</option>
                  <option value="Idowu Samson">Idowu Samson</option>
                  <option value="Mathew Jade">Mathew Jade</option>
                  <option value="Will Smith">Will Smith</option>
                  <option value="Cindy Eze">Cindy Eze</option>
                  <option value="Simon Peter">Simon Peter</option>
                </select>
                        &nbsp;
                <button className="search-sales-by-attendant-buttton link" type="submit">
                  <i className="fas fa-search" />
                  <span className="search-text"> &nbsp; Search</span>
                </button>
              </div>
            </div>
            <div className="dasboard-display-div">
              <form action="" id="add-attendants" onSubmit={e => this.handleSubmit(e)}>
                <div className="form-label fontBlack">
                  <label htmlFor="add-attendants">
                    <i className="far fa-plus-square" />
                  &nbsp;ADD ATTENDANT
                  </label>
                </div>
                <div>
                  <input type="text" name="firstName" id="firstName" placeholder="FIRST NAME" onChange={e => this.onChange(e)} value={firstName} />
                </div>
                <div>
                  <input type="text" name="lastName" id="lastName" placeholder="LAST NAME" onChange={e => this.onChange(e)} value={lastName} />
                </div>
                <div>
                  <input type="email" name="email" id="email" placeholder="EMAIL" onChange={e => this.onChange(e)} value={email} />
                </div>
                <div>
                  <button className="link" id="add-attendant-button" type="submit">CREATE ATTENDANT</button>
                </div>
                {/* { pending ? <Spinner name="circle" className="spinner" id="reactLoader" /> : null} */}
                { pending === true && (
                <Spinner name="circle" className="spinner" id="reactLoader" />
                )}
              </form>
            </div>
          </div>
        </div>
        <ToastContainer />
      </Fragment>
    );
  }
}

AdminAddAttendant.propType = {
  registerAttendant: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired,
  registrationMessage: PropTypes.string.isRequired,
  registrationError: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  registrationMessage: state.registerAttendant.message,
  pending: state.registerAttendant.pending,
  registrationError: state.registerAttendant.error,
});

const mapDispatchToProps = dispatch => ({
  registerAttendant: attendantDetails => dispatch(attendantDetails),
});

export default connect(mapStateToProps, mapDispatchToProps)(AdminAddAttendant);
