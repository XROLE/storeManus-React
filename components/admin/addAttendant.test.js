import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import axios from 'axios';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';
import AddAttendant from './AddAttendant';
import addAttendantReducer from '../../store/reducers/addAttendantReducer';

jest.mock('axios');
describe('Admin Add Attendants', () => {
  it('Should render just the Admin Add Attendant component', () => {
    const component = shallow(
      <AddAttendant />,
    );
    expect(component.exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });
});

describe('Add Component', () => {
  const registerAttendant = jest.fn();
  const initialState = {
    registerAttendant: {
      message: '',
      pending: false,
      error: '',
    },
  };
  const mockStore = configureStore([thunk]);
  const store = mockStore(() => initialState);
  const dispatchMock = jest.spyOn(store, 'dispatch');
  const pending = false;
  const registrationError = '';
  const message = '';

  const props = {
    registerAttendant: {
      registerAttendant,
      pending,
      registrationError,
      message,
    },
  };

  const wrapper = mount(
    <Provider store={store}>
      <BrowserRouter>
        <AddAttendant {...props} />
      </BrowserRouter>
    </Provider>,
  );

  it('should call the regiesterAttendant function', () => {
    wrapper.find('#firstName').simulate(
      'change', {
        target: {
          name: 'firstName', value: 'Harry',
        },
      },
    );

    wrapper.find('#lastName').simulate(
      'change', {
        target: {
          name: 'lastName', value: 'White',
        },
      },
    );

    wrapper.find('#email').simulate(
      'change', {
        target: {
          name: 'email', value: 'harry@white.com',
        },
      },

    );
    axios.post.mockImplementation(() => Promise.resolve({
      data: {
        password: 'jskdfhsjkfgljk;slsa',
      },
    }));
    wrapper.find('form').simulate(
      'submit',
      {
        preventDefault: () => {},
      },
    );
    expect(dispatchMock).toHaveBeenCalled();
  });

  it('should throw error when there is an empty field', () => {
    wrapper.find('#firstName').simulate(
      'change', {
        target: {
          name: 'firstName', value: 'Harry',
        },
      },
    );

    wrapper.find('#lastName').simulate(
      'change', {
        target: {
          name: 'lastName', value: 'White',
        },
      },
    );

    wrapper.find('#email').simulate(
      'change', {
        target: {
          name: 'email', value: '',
        },
      },

    );

    const error = {
      response: {
        data: {
          Message: 'No empty field is allowed',
        },
      },
    };
    axios.post.mockImplementation(() => Promise.reject(error));
    wrapper.find('form').simulate(
      'submit',
      {
        preventDefault: () => {},
      },
    );
    expect(dispatchMock).toHaveBeenCalled();
  });
});

it('should setup default state values', () => {
  const state = addAttendantReducer(undefined, {
    type: '@@INIT',
  });
  expect(state).toEqual({
    message: '',
    pending: false,
    error: '',
  });
});

it('should update state when action type is  ADDING_ATTENDANT', () => {
  const state = addAttendantReducer(undefined, {
    type: 'ADDING_ATTENDANT',
  });
  expect(state).toEqual({
    ...state,
    error: null,
    pending: true,
  });
});

it('should update state when action type is  ADDED_ATTENDANT', () => {
  const state = addAttendantReducer(undefined, {
    type: 'ADDED_ATTENDANT',
    payload: {
      message: 'success',
    },
  });
  expect(state).toEqual({
    ...state,
    message: 'success',
    pending: false,
  });
});

it('should update state when action type is  ADD_ATTENDANT_ERROR', () => {
  const state = addAttendantReducer(undefined, {
    type: 'ADD_ATTENDANT_ERROR',
  });
  expect(state).toEqual({
    ...state,
    pending: false,
  });
});
