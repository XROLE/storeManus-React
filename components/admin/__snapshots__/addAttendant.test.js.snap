// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Add Attendants Should render just the Admin Add Attendant component 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
